csrc/codegen.cpp

```diff
@@ -2972,6 +2960,10 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
           aligned_array_of_regs_.insert(tv);
         }
       }
+      // tv is aligned if alias is aligned
```

Comment:
I'm not sure what we are doing here, not just the change by this PR. When we have an `Allocate` node, does it automatically mean if it's an alias of another tensor but with a different type? It seems that's what is indicated by line 2956 since it does `reinterpret_cast` to an Array type. Is this really safe? Don't we need to check the type of the original allocation of the alias tensor?

Reply:
Looks like that was originally introduced in #665. What do you think, @jacobhinkle?

Reply:
> Is this really safe? Don't we need to check the type of the original allocation of the alias tensor?

Isn't that safe if the two types have the same `sizeof`? That is what #665 is doing.

Reply:
Sorry, my question was incomplete. If the original tensor is defined as a non-aligned tensor, is it safe to reinterpret-cast it to an aligned array type? Isn't that what could happen?

Reply:
That's right if the original tensor had a different alignment from the new tensor. So the danger here would be for example if the original tensor had vectorized access of width 2 then the second tensor which tries to alias it has same sizeof(dtype) but it has a vectorized access of width 8. I didn't think about that in #665 but yes I think we should either guard against that when we're setting up the alias or accomodate the alias vectorized accesses when we codegen the original allocation.

Reply:
Yeah, as far as I can see, the original tensor doesn't even seem to be guaranteed to be aligned at any size.

This doesn't need to be fixed in this PR, but please create an issue.

Reply:
Actually, I forgot about this since I haven't looked at the aliasing code in a while, but we already do this analysis to guarantee the alias vectorizations are at most the same width as the original: https://github.com/NVIDIA/Fuser/blob/d610ce84c25307b9926df1a92e9f928b8e59369c/csrc/device_lower/pass/alias_memory.cpp#L1243-L1247

Reply:
Oh, I see. Glad it'a a false alarm.

Reply:
Actually, I wonder if it should specify the alignment size with `reinterpret_cast`. 

```
<< " = *reinterpret_cast<Array<" << buffer_dtype << ", "
<< genInline(size) << ">*>(&" << genVariableName(alias_tv)
```

If I read this correctly, it just uses an Array type with no alignment requirement, for example, `Array<float, 8>`. The default alignment size is 1, so it seems this would tell the compiler that we are using a non-aligned address with vector loads and stores. The address is indeed aligned, so it should have no problem unless the compiler does something when a given address is marked not aligned properly. I think it'd be safer to use a proper aligned type always even when we know it's properly aligned.

Reply:
Yeah, it's probably not a bad idea to add it: https://github.com/NVIDIA/Fuser/pull/3084